### PR TITLE
refactor: adopt the Core 2026.8 single-owner device registry

### PR DESCRIPTION
**Pull request by jleinenbach**
# Summary

Home Assistant Core 2026.8 turns the device registry into a single-owner model: a device
belongs to exactly one config entry, and the `add_config_entry_id` / `remove_config_entry_id`
keyword pair on `async_update_device` is deprecated. This branch adapts the integration to
that model without raising the declared minimum (`2025.9.1`), by routing every ownership
change through one capability-aware planner and every device lookup through one shared
resolver.

- Type: ☐ fix ☑ feat ☑ refactor ☑ docs ☐ chore ☐ ci ☑ test ☐ perf
- Scope/Area: device registry ownership, coordinator, services, config flow, diagnostics, test harness
- Linked issues: none

## Motivation

Core 2026.8 ships four replacement APIs (`new_config_entry_id`, `new_config_subentry_id`,
`async_get_device_by_identifier`, `async_get_devices`). None of them exists in 2026.7, so
raising the floor to 2026.3 (which was on the table) would not have removed a single
compatibility branch while cutting off users below 2026.3. The floor therefore stays at
`2025.9.1` and every new call site goes through a runtime capability probe instead.

Without this change the integration keeps calling the deprecated keyword pair. On 2026.8
that logs deprecation warnings; the replacement path is the one Core will keep.

---

### Added
- `DeviceRegistryCapabilities` / `detect_device_registry_capabilities`: a runtime probe of
  what the installed Core's device registry supports, so the integration branches on
  measured capability rather than on a version string.
- `OwnershipIntent` / `plan_device_ownership` / `execute_ownership_plan`: call sites now
  state *what* ownership they want, and one planner decides *how* to express it on the
  installed Core.
- `resolve_device_by_identifiers`: one shared resolver for identifier lookups, entry-bound
  on both the modern and the legacy path.
- `device_belongs_to_entry`, `device_owning_entry_ids`, `iter_all_devices`: shared readers,
  because `registry.devices` means different things on the two supported Cores.
- Test harness: a deprecation recorder bound to the `frame` helpers, a single-owner device
  registry double, a differential contract test against the real Core, a ratchet guard
  that fails on any new use of the deprecated keywords, and a guard for the flow-only
  import boundary of `config_flow.py`.
- `script/local_verify.py` now runs an S1..S8 preflight; `script/diff_coverage.py` reports
  patch coverage by intersecting a coverage XML report with the diff against the merge base.
  `make preflight` drives both.

### Changed
- All 35 ownership call sites in `coordinator/registry.py`, plus `services.py`,
  `config_flow.py`, `__init__.py` and `diagnostics.py`, now speak intents instead of registry
  keywords.
- Device counting in diagnostics and the subentry index in `__init__.py` now ask the registry
  per config entry instead of scanning all devices.
- The contract files (`AGENTS.md` at every level, `docs/AI_DEPRECATIONS_GUIDE.md`,
  `docs/CONFIG_SUBENTRIES_HANDBOOK.md`) state the single-owner rule first, so that both
  humans and code assistants see it before touching a call site.

### Fixed
- Subentry index leak: devices owned by a *foreign* config entry could enter this entry's
  subentry index when they carried a legacy identifier. Narrowing the read to
  `dr.async_entries_for_config_entry` closes that.
- A config-flow device lookup used a direct path that is not executable on any supported
  Core version; it now goes through the shared resolver.
- `python script/local_verify.py`, the invocation documented in `script/AGENTS.md`, raised
  `ModuleNotFoundError`, because a direct run puts the script directory on `sys.path` instead
  of the repository root. The tests imported the module rather than starting it, so this was
  never observed.
- A duplicated `-q`: `addopts` already carries one, and the pair suppressed the pytest
  summary line, so preflight logs could not state how many tests had run.
- `script/diff_coverage.py` took its diff from `git diff <merge-base>`, which names tracked
  files only, so a new, unstaged module was listed by `coverage.xml` but never counted as
  changed; the digest clamp in `script/local_verify.py` had the same gap. Both now include
  untracked files (Codex finding on this PR).
- `script/diff_coverage.py` left git's C-style path escapes (`"t\303\274r.py"`) in place, so a
  quoted path could never match its coverage entry; the quoted form is now decoded (Codex
  finding on this PR; no such path exists in this repository today).
- The digest clamp in `script/local_verify.py` read the same git output without `-z` (so a
  quoted path was hashed under its literal, non-existent name) and dropped a path with no file
  behind it (so a deletion during the run compared as unchanged); both git listings now use
  `-z`, and a missing path is recorded as such (Codex findings on this PR).
- `pytest -q` ran as `-qq` everywhere: `[tool.pytest.ini_options] addopts` carried its own
  `-q`, and every documented invocation (`pytest -q` in `AGENTS.md` and `README.md`, the Makefile
  targets, both CI jobs) adds a second, so a red run printed the `FAILED` lines but never the
  summary line that says how many tests failed; the CI failure excerpt (`tail -n 200
  pytest_output.log`) lost exactly that line. `addopts` now carries no `-q`;
  `script/local_verify.py` supplies exactly one (pinned by its test, which had pinned the
  duplication before), and the two Makefile targets that relied on `addopts` carry one now.
  Pre-existing on `main`; the CI workflow is untouched.
- Legacy resolver (found by the pre-push diff review, fixed in the last commit): on Cores
  without `async_get_device_by_identifier`, the shared resolver handed the whole candidate set
  to `async_get_device`, whose `get_entry` returns the first match in *set* iteration order.
  The documented priority therefore held only on the modern branch, and a foreign entry's
  device on the low-priority identifier could shadow this entry's own device on the
  high-priority one. The legacy branch now asks one identifier at a time, in order, and skips
  a foreign hit. Four new tests pin priority and shadowing.
- Three test modules called the 2026.8 registry API on the real registry and would have turned
  the `Test (Python 3.13, declared floor)` job red (15 `AttributeError`s on 2025.9.1). One now
  skips below 2026.8, two go through the shared resolver.
- Flow-only import boundary (Codex, P2): the first cut imported the ownership helpers at module
  scope in `config_flow.py`, which runs `coordinator/__init__.py` and the full API graph on the
  flow-only import path. The import is function-local now, the contract in
  `agents/config_flow/AGENTS.md` says so, and a guard test pins it.
- A dead arm in the service-device sync (`config_subentry_id` fallback under
  `should_remove_service_link`, whose definition requires a non-empty link set) is removed;
  the detach arm and the translation retry next to it, both rewritten by the migration, have
  tests now.

### Removed
- The local device-registry doubles in seven test modules, replaced by the shared
  single-owner double.
- The duplicated identifier translator in `config_flow.py`.

### Performance
- No measured change. Per-entry registry reads replace full-registry scans in three places;
  this was done for correctness, not for speed.

### Security/Privacy
- No change to redaction, token handling or diagnostics payloads. `diagnostics.py` is touched
  in one line (a device count), the redaction path is untouched.

### Compatibility
- No breaking change expected. The declared minimum stays `2025.9.1`; the legacy keyword path
  remains reachable on Cores without the new APIs.
- Verified against three tracks: the locked Core (2026.8.2), a forward-looking one (2026.9.1)
  and the declared floor (2025.9.1). The floor track is the only one on which the legacy
  branch of the planner and the resolver actually runs; it was added after the diff review
  found the resolver bug above.

---

## Tests (MUST)
- ☑ `pytest -q` passes locally
- ☑ New/updated **unit/integration tests** cover the change
- ☑ For **bugfixes**: a **minimal regression test** is included that fails without the fix and passes with it
- Coverage targets:
  - ☐ `config_flow.py` remains **100%** — not met and not newly broken: this repository runs at
    `--cov-fail-under=78` (`ci.yml`, `pyproject.toml`), and the 100%/95% figures in this template
    predate that. Flagged below under Reviewer Notes as a separate follow-up, deliberately not
    changed inside this PR.
  - ☐ Total ≥ **95%** — measured 79.80% after this branch (79.91% on the floor track). See the note above.

### Expected areas (check all that apply)
- ☑ **Config flow**: service device binding via the shared resolver
- ☑ **Lifecycle**: `async_setup_entry`, subentry removal, per-entry device reads
- ☑ **Coordinator & availability**: ownership planning on both capability profiles
- ☑ **Diagnostics**: device count now entry-bound; redaction unchanged
- ☑ **Services**: ownership intents and shared device lookups

---

## Token Cache Safety (MUST)
Not applicable: this PR does not touch token handling, the cache, or any auth path.

---

## Docs & i18n
- ☑ No hard-coded UI strings in Python
- ☐ `translations/*.json` updated — no user-facing strings changed
- ☐ README updated — no user-visible behavior or option changed
- ☑ Contract documents updated (`AGENTS.md` at every level, two `docs/*.md`)

---

## Quality Scale (lightweight, non-blocking)
- `strict-typing`: `mypy --strict` clean, run as stage S3 of the local preflight over the
  file set configured in `pyproject.toml`.
- `test-coverage`: patch coverage 393/401 = 98.00% of changed statements, measured by
  `script/diff_coverage.py` against the post-change coverage report.

---

## Deprecation Check (concise)
- Versions scanned: HA Core 2026.7 through 2026.9.
- Notes found: the 2026.8 single-owner device registry change; `add_config_entry_id` /
  `remove_config_entry_id` on `async_update_device` deprecated in favour of
  `new_config_entry_id` / `new_config_subentry_id`; `registry.devices` no longer a `Mapping`
  from 2026.9.
- Impact here: all four replacement APIs are used behind a runtime capability probe; a guard
  test fails on any new use of the deprecated keywords.

---

## Risk & Rollback
- Risk level: ☐ low ☑ medium ☐ high — wide blast radius (62 files), but behaviour-preserving
  on the declared minimum and covered by a differential contract test against the real Core.
- Rollback: revert the branch. No data migration, no entity ID change, no stored state touched.

---

## Local Verification (run before pushing)

Run as `make preflight`, which drives `script/local_verify.py --all` over the interpreters
named in `PREFLIGHT_PYTHONS`. Result of the final run over three interpreters (exit 0; S7/S8
are reported NOT CHECKED with an expected reason, which the script does not count as a
failure, see the `expected_absent` note in `StageResult`):

| Stage | Result | Detail |
|---|---|---|
| S1 format (`ruff format --check`) | OK | |
| S2 lint (`ruff check`) | OK | |
| S3 types (`mypy --strict`) | OK | over the file set configured in `pyproject.toml` |
| S4 spelling (`codespell`) | NOTE | 90 findings tree-wide, 22 on the 62 files of this branch; the CI step carries `continue-on-error` |
| S5 suite and project coverage | OK | 7269 tests on each track, zero failures: 7244 passed / 25 skipped on Core 2026.8.2, 7250 / 19 on Core 2026.9.1, 7227 / 42 on Core 2025.9.1; project coverage 79.80% / 79.80% / 79.91% against a gate of 78% |
| S6 patch coverage | OK | 393/401 changed statements covered = 98.00%, against a threshold of 82.21% |
| S7 security scans (`bandit`, `semgrep`, `pip-audit`) | NOT CHECKED | tooling not available locally |
| S8 manifest (`hassfest`) | NOT CHECKED | requires Docker, not available locally |

**S7 and S8 have not run.** A stage that could not run is not a stage that passed. Those two
are seeing this change for the first time in CI, and that is stated here rather than left for
a reviewer to discover.

---

## Reviewer Notes

* **Why one PR and not eighteen.** The work is one migration with one invariant, and the
  intermediate states are not independently shippable: the guard ratchet in the test phase
  fails until the production call sites follow. The branch is therefore cut into 20 commits,
  one per work package, one for the findings of the pre-push diff review, one per
  bot round and one for a pre-existing `-qq` in every pytest invocation, each with its own
  rationale in the commit body. Reviewing commit by
  commit should be easier than reviewing the squashed diff.
* **Codecov's 22 lines, accounted for one by one** (commit 16 body has the detail): 15
  changed statements without a hit, every one a rewrite of a line that was already uncovered
  on the base commit, plus 7 partials. Two are covered now, one was dead and is gone, two
  arms are excluded with a measured reason, five in `services.py` are pre-existing and were
  measured on the base tree, three in `__init__.py` sit in a function that has no caller.
  The three `exec()` findings from Semgrep are gone as well: none of the three test helpers
  needed `exec`.
* **Follow-up, deliberately not in this PR:** this template asks for 95% total and 100% on
  `config_flow.py`, while the repository gates at `--cov-fail-under=78`. That contradiction
  predates this branch. Lowering a coverage promise inside the very PR that adds test code is
  the wrong move; it belongs in its own change.
* **Follow-up, deliberately not in this PR:** `Makefile` (two targets) and
  `.github/workflows/ci.yml` pass `-q` on top of the one in `addopts`, which suppresses the
  pytest summary line. In CI this matters, because the failure excerpt (`tail -n 200`) then
  cannot state how many tests failed. Fixing it means opening the CI definition, which does
  not belong in a migration PR.
* **Follow-up, deliberately not in this PR:**
  `tests/test_docker_login_hardening.py::test_wait_helper_behaviourally_survives_a_child_that_outlives_one_signal`
  failed once on the floor track under a load average of about 7.7 on four cores and passed on
  the repeated run under the same load; isolated, it fails in 2 of 8 runs under that load and
  0 of 5 without it. It is unrelated to this diff: the test sends a signal after a fixed
  `sleep(0.5)` to a child that installs its handler only after start, so there is no
  handshake. The helper under test behaves correctly. Noted here so that a red run of that
  test in CI is read as what it is.
* Edge cases to double-check: the legacy branch of `plan_device_ownership` (only reachable on
  Cores without the 2026.8 APIs), and the subentry index narrowing in `__init__.py`.

### CODE REVIEW

**Comment by jleinenbach** [comment]
@codex review

**Comment by jleinenbach** [comment]
@codex review

**Comment by jleinenbach** [comment]
@codex review

File: tests/helpers/deprecation_recorder.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+
+    Deprecated Core APIs branch on *who* called them: from a custom integration
+    they log, from anywhere else they raise.  Tests that want the logging branch
+    therefore need a frame below ``custom_components/googlefindmy/``.  Compiling
+    a trampoline with that file name is the least invasive way to get one; the
+    alternative would be adding test-only helpers to production code.
+    """
+    source = "def _invoke(func, args, kwargs):\n    return func(*args, **kwargs)\n"
+    namespace: dict[str, Any] = {}
+    exec(compile(source, PROBE_FILENAME, "exec"), namespace)  # noqa: S102
```
[the remark this answers is not in the document]
Comment (jleinenbach) [reply tests/helpers/deprecation_recorder.py]: Fixed in a77fc382: the trampoline is instantiated from the compiled code object (FunctionType over the CodeType in co_consts), so the frame keeps PROBE_FILENAME as co_filename and no module-level code is executed. No exec left in this helper.

File: tests/test_coordinator_helpers_registry.py [hunk 2]
```diff
@@ -1028,3 +1065,1173 @@
...
+    bool(_missing_symbols),
+    reason=f"AP-10 has not landed yet; missing: {', '.join(_missing_symbols)}",
+)
+
+
+def _profile(**parameters: object) -> object:
+    """Build a callable whose signature carries exactly ``parameters``."""
+    names = ", ".join(f"{name}=None" for name in parameters)
+    namespace: dict[str, object] = {}
+    exec(f"def _call(device_id, *, {names}):\n    return None\n", namespace)  # noqa: S102
```
[the remark this answers is not in the document]
Comment (jleinenbach) [reply tests/test_coordinator_helpers_registry.py]: Fixed in a77fc382: _profile declares its keyword-only parameters through __signature__, which is what detect_device_registry_capabilities reads via inspect.signature. No exec.

File: tests/test_coordinator_helpers_registry.py [hunk 3]
```diff
@@ -1028,3 +1065,1173 @@
...
+
+    def test_var_keyword_fallback(self) -> None:
+        """A ``**kwargs`` signature accepts anything, so assume the legacy name.
+
+        Test doubles are the realistic case here.  Guessing "single owner"
+        instead would send ``new_config_subentry_id`` at a double that silently
+        swallows it, and the test would pass while production broke.
+        """
+        namespace: dict[str, object] = {}
+        exec("def _call(device_id, **kwargs):\n    return None\n", namespace)  # noqa: S102
```
[the remark this answers is not in the document]
Comment (jleinenbach) [reply tests/test_coordinator_helpers_registry.py]: Fixed in a77fc382: a plain def _call(device_id, **kwargs) does the job; exec was never needed here.

File: custom_components/googlefindmy/config_flow.py [hunk 4]
```diff
@@ -172,6 +172,13 @@
     coerce_ignored_mapping,
     service_device_identifier,
 )
+from .coordinator.helpers.registry import (
```
[the remark this answers is not in the document]
Comment (jleinenbach) [reply custom_components/googlefindmy/config_flow.py]: Fixed in a77fc382: the ownership helpers are imported inside _ensure_service_device_binding (the one function that uses them), agents/config_flow/AGENTS.md states the rule, and tests/test_config_flow_lazy_registry_imports.py pins it (module attributes absent, no module-scope import of .coordinator or .api in the AST, with a positive control for the AST walk).

File: script/diff_coverage.py [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+
+    base = _run_git(["merge-base", base_ref, "HEAD"], repo_root).strip()
+    if not base:
+        raise MeasurementError(f"no merge base between {base_ref} and HEAD")
+    # The working tree, not `base..HEAD`. The coverage report this is intersected
+    # with comes from a run over the current tree, so a diff that stops at HEAD
+    # would describe a different revision: uncommitted lines would be missing
+    # from the diff while their line numbers had already shifted the report.
+    # Two sides of one intersection have to mean the same revision.
+    diff = _run_git(["diff", "--unified=0", base], repo_root)
```
[the remark this answers is not in the document]
Comment (jleinenbach) [reply script/diff_coverage.py:408]: Fixed in 82afda91: collect_diff appends one synthetic section per untracked file (git ls-files --others --exclude-standard -z), an unreadable listed file is a MeasurementError, and the same gap in local_verify's digest clamp (_changed_paths from git diff HEAD only) is closed as well. Tests cover the 1/2 case and the vanished-file case; the finding was real here (the previous preflight ran with one untracked test module).

File: script/diff_coverage.py [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+
+def _diff_path(target: str) -> str | None:
+    """Return the repository-relative path a diff header names, if it names one."""
+
+    if target == "/dev/null":
+        return None
+    # git quotes paths that contain spaces or non-ASCII bytes; the quotes are not
+    # part of the name and would never match a coverage entry.
+    if target.startswith('"') and target.endswith('"'):
+        target = target[1:-1]
```
[the remark this answers is not in the document]
Comment (jleinenbach) [reply script/diff_coverage.py]: Fixed in 0cf6f8f6: _unquote_git_path decodes git's C-style quoting (the C escapes plus three-digit octal bytes) into UTF-8; a truncated or unknown escape is a MeasurementError. Probed against a real `git diff --unified=0 HEAD` over a path with a non-ASCII byte and an escaped quote. No such path exists in this repository today, so the branch numbers are unchanged.

**Comment by jleinenbach** [comment]
@codex review

**Comment by jleinenbach** [comment]
@codex review

File: script/local_verify.py [hunk 7]
```diff
@@ -60,8 +139,411 @@ def _build_pytest_command(pytest_args: Sequence[str] | None) -> list[str]:
...
+    tree and not yet staged is part of the tree the suite runs on, so the digest
+    clamp has to hold it as well; otherwise an edit to a new file during the
+    run would go unnoticed, and the clamp would report one file fewer than the
+    tree actually changed.
+    """
+
+    paths: list[str] = []
+    for command in (
+        ["git", "diff", "HEAD", "--name-only"],
+        ["git", "ls-files", "--others", "--exclude-standard"],
```
[the remark this answers is not in the document]
Comment (jleinenbach) [reply script/local_verify.py]: Fixed in 0f050497: `_changed_paths` and `_branch_paths` now read all three git listings with `-z` and split on NUL (`_nul_separated`), so git hands the name over verbatim and nothing needs decoding before hashing. Covered by `test_git_paths_arrive_verbatim_under_the_default_quoting`, which commits `tür "x".py` into a throwaway repository with `core.quotePath=true` and includes a positive control that git does quote there.

File: script/local_verify.py [hunk 8]
```diff
@@ -60,8 +139,411 @@ def _build_pytest_command(pytest_args: Sequence[str] | None) -> list[str]:
...
+
+
+def _digest_of(paths: Sequence[str]) -> dict[str, str]:
+    """Return a SHA-256 digest per existing path, relative to the repository."""
+
+    digests: dict[str, str] = {}
+    for name in paths:
+        candidate = REPO_ROOT / name
+        if candidate.is_file():
+            digests[name] = hashlib.sha256(candidate.read_bytes()).hexdigest()
```
[the remark this answers is not in the document]
Comment (jleinenbach) [reply script/local_verify.py:187]: Fixed in 0f050497: `_digest_of` records a path with no file behind it as `MISSING_DIGEST` instead of leaving it out, so a file deleted during the run differs between the two snapshots and trips the clamp. Covered by `test_a_missing_path_is_recorded_not_dropped` and `test_a_deletion_during_the_run_discards_it`.

**Comment by jleinenbach** [comment]
@codex review
